src/bun.js/bindings/VMModuleScript.cpp

```diff
@@ -0,0 +1,277 @@
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMError(globalObject, scope, "TODO"_s);
+}
+
+JSC_DEFINE_HOST_FUNCTION(scriptRunInContext, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+
+    JSValue thisValue = callFrame->thisValue();
+    auto* script = jsDynamicCast<VMModuleScript*>(thisValue);
+    if (UNLIKELY(!script)) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "Script.prototype.runInContext can only be called on a Script object"_s);
+    }
+
+    ArgList args(callFrame);
+
+    JSValue contextArg = args.at(0);
+    if (!UNLIKELY(contextArg.isObject())) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "context parameter must be a contextified object"_s);
+    }
+    JSObject* context = asObject(contextArg);
+
+    auto* zigGlobalObject = reinterpret_cast<Zig::GlobalObject*>(globalObject);
+    JSValue scopeVal = zigGlobalObject->vmModuleContextMap()->get(context);
+    if (UNLIKELY(scopeVal.isUndefined())) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "context parameter must be a contextified object"_s);
+    }
+    JSScope* scope = jsDynamicCast<JSScope*>(scopeVal);
+    ASSERT(scope);
+
+    return runInContext(globalObject, script, context, scope, args.at(1));
+}
+JSC_DEFINE_HOST_FUNCTION(scriptRunInThisContext, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    JSValue thisValue = callFrame->thisValue();
+    auto* script = jsDynamicCast<VMModuleScript*>(thisValue);
+    if (UNLIKELY(!script)) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "Script.prototype.runInThisContext can only be called on a Script object"_s);
+    }
+
+    ArgList args(callFrame);
+    return runInContext(globalObject, script, globalObject->globalThis(), globalObject->globalScope(), args.at(0));
+}
+
+JSC_DEFINE_CUSTOM_GETTER(scriptGetSourceMapURL, (JSGlobalObject* globalObject, EncodedJSValue thisValue, PropertyName))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMError(globalObject, scope, "TODO"_s);
+}
+
+JSC_DEFINE_HOST_FUNCTION(vmModule_createContext, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    ArgList args(callFrame);
+    JSValue contextArg = args.at(0);
+    if (!contextArg.isObject()) {
+        return throwVMTypeError(globalObject, scope, "parameter to createContext must be an object"_s);
+    }
+    JSObject* context = asObject(contextArg);
+
+    JSScope* contextScope = JSWithScope::create(vm, globalObject, globalObject->globalScope(), context);
```

Comment:
I think we have to use a new `JSGlobalObject` since if you assign to a global in a context it shouldn't impact the calling context
<img width="491" alt="image" src="https://user-images.githubusercontent.com/709451/236962779-02386818-2fc3-4975-adfb-2de140f54ebe.png">

I think we can use JSProxy here:

https://github.com/oven-sh/WebKit/blob/6bd997e38e30f50b3114b178f4d5e2fc5da6da25/Source/WebCore/bindings/js/JSDOMGlobalObject.cpp#L699-L707

WebCore has to do some similar strange things due to the combo of:
- Web Extensions
- iframes accessing parent window
- regular userland code